compiler/rustc_const_eval/src/interpret/machine.rs

```diff
@@ -480,17 +496,28 @@ pub macro compile_time_machine(<$mir: lifetime, $tcx: lifetime>) {
     }
 
     #[inline(always)]
-    fn ptr_from_addr(_ecx: &InterpCx<$mir, $tcx, Self>, addr: u64) -> Pointer<Option<AllocId>> {
+    fn ptr_from_addr_transmute(
+        _ecx: &InterpCx<$mir, $tcx, Self>,
+        addr: u64,
+    ) -> Pointer<Option<AllocId>> {
+        Pointer::new(None, Size::from_bytes(addr))
+    }
+
+    #[inline(always)]
+    fn ptr_from_addr_cast(
+        _ecx: &InterpCx<$mir, $tcx, Self>,
+        addr: u64,
+    ) -> Pointer<Option<AllocId>> {
         Pointer::new(None, Size::from_bytes(addr))
```

Comment:
This should also error in CTFE and ConstProp.

Reply:
```rust
const fn foo() {
    let a = 5 as *const i32;
}
```

I don't think we can do that, because the above already compiles on stable. Did you mean `ptr_from_addr_transmute`?

Reply:
Oh, I forgot that we allow this. *oops*

The transmute can also be triggered on stable I think, and there is nothing fundamentally wrong with transmuting ints to ptrs. So yeah let's just make both of them return `None` provenance for now. Since no pointer is ever exposed, this is actually correct.  :)